### PR TITLE
Various Redox OS fixes and add i686 Redox OS target

### PR DESCRIPTION
**Pull request by bjorn3**
All of these come from the fork used by Redox OS available at https://gitlab.redox-os.org/redox-os/rust/-/commits/redox-2024-05-11/?ref_type=heads.

cc @jackpot51

### CODE REVIEW

**Comment by bjorn3** [comment]
> A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

@jackpot51

> Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

The x86_64 and aarch64 redox targets are named *-unknown-redox too.

> Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.

Redox OS is permissively licensed and uses a regular C toolchain.

> Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.

Agreed

> Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

Support should be identical to the existing x86_64 and aarch64 targets.

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running binaries, or running tests (even if they do not pass), the documentation must explain how to run such binaries or tests for the target, using emulation if possible or dedicated hardware if necessary.

Same as x86_64 redox os.

> Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.

Agreed

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.

Agreed

> Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target.

It does.

**Review by jackpot51 — APPROVED** [review]

File: src/doc/rustc/src/platform-support.md [hunk 1]
```diff
@@ -300,6 +300,7 @@ target | std | host | notes
 [`i686-unknown-hurd-gnu`](platform-support/hurd.md) | ✓ | ✓ | 32-bit GNU/Hurd [^x86_32-floats-return-ABI]
 [`i686-unknown-netbsd`](platform-support/netbsd.md) | ✓ | ✓ | NetBSD/i386 with SSE2 [^x86_32-floats-return-ABI]
 [`i686-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 32-bit OpenBSD [^x86_32-floats-return-ABI]
+`i686-unknown-redox` | ? |  | i686 Redox OS
```
Comment (Urgau) [inline src/doc/rustc/src/platform-support.md:303]: I believe every new target **must** have a dedicated platform support page.

Could you add one for this target (or maybe one for all Redox targets) by adding a page under the `platform-support` subdirectory[^1] and link it here as done for other targets.

[^1]: Don't forget to also add the newly created file in the [`SUMMARY.md`](https://github.com/rust-lang/rust/blob/db45105694cc5a63d1c1b39ebb725a024cea2d13/src/doc/rustc/src/SUMMARY.md) file.

Comment (bjorn3) [reply src/doc/rustc/src/platform-support.md]: Added

**Comment by fee1-dead** [comment]
r? compiler

**Comment by bjorn3** [comment]
Fixed a trivial merge conflict in the platform support page.

@bors r=petrochenkov

**Comment by bjorn3** [comment]
I don't know how the target sanity check worked previously with newly added targets (see https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/STAGE0_MISSING_TARGETS.20broken), but it should be fixed with the latest commit by skipping the target sanity check when doing a local rebuild.

**Comment by bjorn3** [comment]
Seems like skipping local rebuilds doesn't fix it. I've removed i686-unknown-redox from STAGE0_MISSING_TARGETS for now. Turns out adding new targets here isn't actually necessary.

File: compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    base.plt_by_default = false;
+    base.max_atomic_width = Some(64);
+    base.add_pre_link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-m32"]);
+    // don't use probe-stack=inline-asm until rust#83139 and rust#84667 are resolved
+    base.stack_probes = StackProbeType::Call;
+
+    Target {
+        llvm_target: "i686-unknown-redox".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: None,
```
Comment (L3P3) [inline compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs:15]: Why no metadata?

Comment (bjorn3) [reply compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs:15]: The x86_64 target didn't have it either at least when I made this PR.

File: compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+use crate::spec::{base, Cc, LinkerFlavor, Lld, StackProbeType, Target};
+
+pub fn target() -> Target {
+    let mut base = base::redox::opts();
+    base.cpu = "pentiumpro".into();
```
Comment (RalfJung) [inline compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs:5]: This makes i686_unknown_redox our only "i686" target that does not use pentium4 as baseline. Is that deliberate?  It is certainly very surprising.

File: src/doc/rustc/src/platform-support.md [hunk 4]
```diff
@@ -300,6 +300,7 @@ target | std | host | notes
 [`i686-unknown-hurd-gnu`](platform-support/hurd.md) | ✓ | ✓ | 32-bit GNU/Hurd [^x86_32-floats-return-ABI]
 [`i686-unknown-netbsd`](platform-support/netbsd.md) | ✓ | ✓ | NetBSD/i386 with SSE2 [^x86_32-floats-return-ABI]
 [`i686-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 32-bit OpenBSD [^x86_32-floats-return-ABI]
+[`i686-unknown-redox`](platform-support/redox.md) | ✓ |  | i686 Redox OS
```
Comment (RalfJung) [inline src/doc/rustc/src/platform-support.md:303]: When adding an x86-32 target, please always add the appropriate footnote for the float issues on that target.

File: compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs (hunk 3 above)
Comment (RalfJung) [reply compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs:5]: Ah no it is not alone, just some of the footnotes were wrong...

Comment (RalfJung) [reply compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs:5]: So in total there are 2 i686 targets that do not use pentium4 as a baseline: this one and i686-unknown-hurd-gnu. I still think that's a bug, or was it explicitly discussed somewhere to deviate from our usual naming scheme? (And yes I know that naming scheme is technically wrong but we're not renaming a tier 1 target.)

**Pull request merged**